compiler/rustc_codegen_ssa/src/back/symbol_export.rs

```diff
@@ -321,6 +321,31 @@ fn exported_symbols_provider_local(
 
         let cgus = tcx.collect_and_partition_mono_items(()).codegen_units;
 
+        // Do not export symbols that cannot be instantiated by downstream crates.
+        let reachable_set = tcx.reachable_set(());
+        let is_local_to_current_crate = |ty: Ty<'_>| {
+            let no_refs = ty.peel_refs();
+            let root_def_id = match no_refs.kind() {
+                rustc_middle::ty::Closure(closure, _) => *closure,
+                rustc_middle::ty::FnDef(def_id, _) => *def_id,
+                rustc_middle::ty::Coroutine(def_id, _) => *def_id,
+                rustc_middle::ty::CoroutineClosure(def_id, _) => *def_id,
+                rustc_middle::ty::CoroutineWitness(def_id, _) => *def_id,
+                _ => return false,
+            };
+            let Some(root_def_id) = root_def_id.as_local() else {
+                return false;
+            };
+
+            let is_local = !reachable_set.contains(&root_def_id);
+            is_local
+        };
+
+        let is_instantiable_downstream = |did, type_args| {
+            std::iter::once(tcx.type_of(did).skip_binder()).chain(type_args).all(|arg| {
```

Comment:
Maybe putting `type_args` before `type_of(did)` would be faster by allowing short-circuiting of the `type_of` query when `type_args` already contains a local type?